stdlib/public/core/StringUnicodeScalarView.swift

```diff
@@ -186,52 +159,43 @@ extension String {
     ///   must be less than the view's end index.
     @_inlineable // FIXME(sil-serialize-all)
     public subscript(position: Index) -> Unicode.Scalar {
-      var scratch = _ScratchIterator(_guts, _toCoreIndex(position))
-      var decoder = UTF16()
-      switch decoder.decode(&scratch) {
+      switch _guts._decodeUnicodeScalar(startingAt: _toCoreIndex(position)) {
       case .scalarValue(let us):
         return us
       case .emptyInput:
-        _sanityCheckFailure("cannot subscript using an endIndex")
+        _preconditionFailure("cannot subscript using an endIndex")
       case .error:
-        return Unicode.Scalar(0xfffd)!
+        return Unicode.Scalar._replacementCharacter
       }
     }
 
     /// An iterator over the Unicode scalars that make up a `UnicodeScalarView`
     /// collection.
     @_fixed_layout // FIXME(sil-serialize-all)
     public struct Iterator : IteratorProtocol {
-      @_inlineable // FIXME(sil-serialize-all)
       @_versioned // FIXME(sil-serialize-all)
-      internal init(_ _base: _LegacyStringCore) {
-        self._iterator = _base.makeIterator()
-        if _base.hasContiguousStorage {
-          self._baseSet = true
-          if _base.isASCII {
-            self._ascii = true
-            self._asciiBase = UnsafeBufferPointer(
-              start: _base._baseAddress?.assumingMemoryBound(
-                to: UTF8.CodeUnit.self),
-              count: _base.count).makeIterator()
-          } else {
-            self._ascii = false
-            self._base = UnsafeBufferPointer<UInt16>(
-              start: _base._baseAddress?.assumingMemoryBound(
-                to: UTF16.CodeUnit.self),
-              count: _base.count).makeIterator()
-          }
-        } else {
-          self._ascii = false
-          self._baseSet = false
-        }
-      }
+      internal var _guts: _StringGuts
+
+      @_versioned // FIXME(sil-serialize-all)
+      internal var _asciiIterator: _UnmanagedASCIIString.UnicodeScalarIterator?
+      @_versioned // FIXME(sil-serialize-all)
+      internal var _utf16Iterator: _UnmanagedUTF16String.UnicodeScalarIterator?
+      @_versioned // FIXME(sil-serialize-all)
+      internal var _opaqueIterator: _UnmanagedOpaqueString.UnicodeScalarIterator?
```

Comment:
Something that might be interesting is what we want our view's ABI to be. Here, we almost certainly want them unified in an enum-like representation to save space. But, that's pretty similar to me how that is different than `_StringGuts` itself and an index. I assume storing these is to avoid the _StringGuts introspection and associated ARC?

Reply:
Exactly! I tried doing it with just an index, but constant introspection killed performance. I also hate having to waste space on three iterators though.

Reply:
We need to speed up the perf of introspection. That's something I can try to hand-tune again (possibly changing bit-representations), but probably won't get around to it for a few days.